### PR TITLE
Create ldap identity plugin #513

### PR DESCRIPTION
**Pull request by paulofelipefeitosa**
PR associated with task [#513](https://git.lsd.ufcg.edu.br/chicog/fogbow/issues/513).

### CODE REVIEW

File: src/main/java/org/fogbowcloud/manager/core/models/token/Token.java [hunk 1]
```diff
@@ -1,5 +1,189 @@
...
+import java.util.Map;
+
+import org.fogbowcloud.manager.core.utils.DateUtils;
+import org.fogbowcloud.manager.core.utils.JSONHelper;
+import org.json.JSONException;
+import org.json.JSONObject;
+
 public class Token {
+	
+	private static final String DATE_EXPIRATION = "dateExpiration";
```
Comment (llvieira) [inline src/main/java/org/fogbowcloud/manager/core/models/token/Token.java]: Rename to `EXPIRATION_DATE = "expirationDate"`.

**Comment by thiagomanel** [comment]
There is a typo in this code (An "Erro" instead or "Error")

File: src/main/java/org/fogbowcloud/manager/core/plugins/identity/ldap/LdapIdentityPlugin.java [hunk 2]
```diff
@@ -0,0 +1,344 @@
...
+		this.ldapUrl = properties.getProperty(PROP_LDAP_URL);
+		this.encryptType = properties.getProperty(PROP_LDAP_ENCRYPT_TYPE);
+		this.privateKeyPath = properties.getProperty(PRIVATE_KEY_PATH);
+		this.publicKeyPath = properties.getProperty(PUBLIC_KEY_PATH);
+	}
+
+	@Override
+	public Token createToken(Map<String, String> userCredentials) {
+
+		String uid = userCredentials.get(CRED_USERNAME);
```
Comment (llvieira) [inline src/main/java/org/fogbowcloud/manager/core/plugins/identity/ldap/LdapIdentityPlugin.java]: Change to `userId` or `uuid`.

File: src/main/java/org/fogbowcloud/manager/core/models/Credential.java [hunk 3]
```diff
@@ -1,5 +1,33 @@
...
+	}
+	
+	public void setValueDefault(String valueDefault) {
+		this.valueDefault = valueDefault;
+	}
+
+	public boolean isRequired() {
+		return required;
+	}
 }
```
Comment (thiagomanel) [inline src/main/java/org/fogbowcloud/manager/core/models/Credential.java]: I did not check this code out to compile it. However, do we really need this set methods? Could not this class be immutable

File: src/main/java/org/fogbowcloud/manager/core/plugins/identity/ldap/LdapIdentityPlugin.java [hunk 4]
```diff
@@ -0,0 +1,344 @@
...
+	}
+
+	protected boolean verifySign(String tokenMessage, String signature) {
+		RSAPublicKey publicKey = null;
+		try {
+			publicKey = RSAUtils.getPublicKey(this.publicKeyPath);
+			return RSAUtils.verify(publicKey, tokenMessage, signature);
+		} catch (Exception e) {
+			LOGGER.error("Erro while trying to validate sing of the token.", e);
+			throw new RuntimeException("Erro while trying to validate sing of the token.");
```
Comment (llvieira) [inline src/main/java/org/fogbowcloud/manager/core/plugins/identity/ldap/LdapIdentityPlugin.java]: Same as above.

File: src/main/java/org/fogbowcloud/manager/core/models/token/Token.java [hunk 5]
```diff
@@ -1,5 +1,189 @@
...
+			return true;
+		}
 
+		@Override
+		public String toString() {
+			return "User [id=" + id + ", name=" + name + "]";
+		}				
+		
+	}
+
```
[the remark this answers is not in the document]
Comment (paulofelipefeitosa) [reply src/main/java/org/fogbowcloud/manager/core/models/token/Token.java]: @thiagomanel I don't know, by the fact that the authentication plugin isn't already implemented. All this methods are automatic generated by the IDE, do you think that is necessary change it?

File: src/test/java/org/fogbowcloud/manager/core/plugins/PluginHelper.java [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+	public static final String USER_ID = "user_id";
+	public static final String USERNAME = "admin";
+	public static final String USER_PASS = "reverse";
+
+	public static final String FED_USERNAME = "federation_user";
+	public static final String FED_USER_PASS = "federation_user_pass";
+	
+	/**
+	 * Getting a available port on range 60000:61000
```
[the remark this answers is not in the document]
Comment (paulofelipefeitosa) [reply src/test/java/org/fogbowcloud/manager/core/plugins/PluginHelper.java]: @thiagomanel [here](https://github.com/fogbow/fogbow-manager-core/blob/4311be718acb243c8bf0a129088160a0bbe5ca22/src/test/java/org/fogbowcloud/manager/core/plugins/identity/ldap/TestLdapIdentityPlugin.java#L21)

File: src/main/java/org/fogbowcloud/manager/core/models/token/Token.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (thiagomanel) [reply src/main/java/org/fogbowcloud/manager/core/models/token/Token.java]: sure. If we can keep only the GETs, do so

File: src/test/java/org/fogbowcloud/manager/core/plugins/PluginHelper.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (thiagomanel) [reply src/test/java/org/fogbowcloud/manager/core/plugins/PluginHelper.java]: I don't like the idea of creating a public method only to be used on test code. Why not moving this to a private method within the very test class?

File: src/main/java/org/fogbowcloud/manager/core/models/token/Token.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (paulofelipefeitosa) [reply src/main/java/org/fogbowcloud/manager/core/models/token/Token.java]: @thiagomanel by the way, which methods do you talking about?

[the remark this answers is not in the document]
Comment (thiagomanel) [reply src/main/java/org/fogbowcloud/manager/core/models/token/Token.java]: all method named as setSOMETHING

File: src/main/java/org/fogbowcloud/manager/core/models/Credential.java (hunk 3 above)
Comment (paulofelipefeitosa) [reply src/main/java/org/fogbowcloud/manager/core/models/Credential.java]: Done! Check [this](https://github.com/fogbow/fogbow-manager-core/commit/f091c029cb9afbf263381d4112277e16e340e449) out.

File: src/main/java/org/fogbowcloud/manager/core/plugins/identity/ldap/LdapIdentityPlugin.java (hunk 4 above)
Comment (paulofelipefeitosa) [reply src/main/java/org/fogbowcloud/manager/core/plugins/identity/ldap/LdapIdentityPlugin.java]: Done :). Check [this](https://github.com/fogbow/fogbow-manager-core/commit/78864ce8eefde35b7118f3caa28d205a7ff8ce8d) out.

File: src/main/java/org/fogbowcloud/manager/core/plugins/identity/ldap/LdapIdentityPlugin.java [hunk 7]
```diff
@@ -0,0 +1,344 @@
...
+		return signature;
+	}
+
+	protected boolean verifySign(String tokenMessage, String signature) {
+		RSAPublicKey publicKey = null;
+		try {
+			publicKey = RSAUtils.getPublicKey(this.publicKeyPath);
+			return RSAUtils.verify(publicKey, tokenMessage, signature);
+		} catch (Exception e) {
+			LOGGER.error("Erro while trying to validate sing of the token.", e);
```
[the remark this answers is not in the document]
Comment (paulofelipefeitosa) [reply src/main/java/org/fogbowcloud/manager/core/plugins/identity/ldap/LdapIdentityPlugin.java]: Done :). Check [this](https://github.com/fogbow/fogbow-manager-core/commit/78864ce8eefde35b7118f3caa28d205a7ff8ce8d) out.

File: src/main/java/org/fogbowcloud/manager/core/plugins/identity/ldap/LdapIdentityPlugin.java (hunk 2 above)
Comment (paulofelipefeitosa) [reply src/main/java/org/fogbowcloud/manager/core/plugins/identity/ldap/LdapIdentityPlugin.java]: Done :). Check [this](https://github.com/fogbow/fogbow-manager-core/commit/78864ce8eefde35b7118f3caa28d205a7ff8ce8d) out.

File: src/main/java/org/fogbowcloud/manager/core/models/token/Token.java (hunk 1 above)
Comment (paulofelipefeitosa) [reply src/main/java/org/fogbowcloud/manager/core/models/token/Token.java]: Done :). Check [this](https://github.com/fogbow/fogbow-manager-core/commit/8aba75e25659c73671a6bbc0feac76d351d6a70f) out.

File: src/main/java/org/fogbowcloud/manager/core/utils/DateUtils.java [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+import java.text.SimpleDateFormat;
+import java.util.Date;
+import java.util.Locale;
+import java.util.TimeZone;
+
+public class DateUtils {
+
+	public static final String ISO_8601_DATE_FORMAT = "yyyy-MM-dd'T'HH:mm:ss.SSSZ";
+
+	public DateUtils() {
```
[the remark this answers is not in the document]
Comment (paulofelipefeitosa) [reply src/main/java/org/fogbowcloud/manager/core/utils/DateUtils.java]: Done :). Check [this](https://github.com/fogbow/fogbow-manager-core/commit/78864ce8eefde35b7118f3caa28d205a7ff8ce8d) out.

File: src/main/java/org/fogbowcloud/manager/core/plugins/identity/ldap/LdapIdentityPlugin.java [hunk 9]
```diff
@@ -0,0 +1,344 @@
...
+
+	@Override
+	public String getAuthenticationURI() {
+		return null;
+	}
+
+	@Override
+	public Token getForwardableToken(Token originalToken) {
+		return originalToken;
+	}
```
[the remark this answers is not in the document]
Comment (paulofelipefeitosa) [reply src/main/java/org/fogbowcloud/manager/core/plugins/identity/ldap/LdapIdentityPlugin.java]: Done :). Check [this](https://github.com/fogbow/fogbow-manager-core/commit/78864ce8eefde35b7118f3caa28d205a7ff8ce8d) out.

File: src/main/java/org/fogbowcloud/manager/core/models/token/Token.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (paulofelipefeitosa) [reply src/main/java/org/fogbowcloud/manager/core/models/token/Token.java]: @thiagomanel setDateUtils is necessary, check [this](https://github.com/fogbow/fogbow-manager-core/pull/8/commits/7da256c2497766f6ffe351acfb791b7164c9bb6c) out.

**Comment by paulofelipefeitosa** [comment]
@thiagomanel please, review again with the modifications that you requested and merge this pull request. :)

**Pull request merged**